### DISCUSSION TITLE
Improving the github.io interface

### DISCUSSION THREAD

**Discussion by arunchawla-NOAA** [NOAA-EMC/NOAA-EMC.github.io#17] [General]
Can we update this ? I cannot see the link to the NCEPLIBS-bufr documentation and there are bullets to things that have no documentation. Plus when I follow the path I get the directories and not the html page links

https://noaa-emc.github.io/

**Comment by arunchawla-NOAA** [comment]
This link is at the top of our repositories page. Can we please clean this up ? It does not look good at all

**Comment by aerorahul** [comment]
This repository needs to be cleaned up.  The documentation uploaded here should be removed.  It does not belong in this repository.

**Comment by arunchawla-NOAA** [comment]
are all the doxygen html files uploaded here ? I have put a number of people here just for discussion. Final task will be assigned later. I just want to get a scope of what is happening here

**Reply by edwardhartnett** [reply]
This is a legacy method to distribute documentation. Using GitHub gh-pages directly provides better results, as mentioned by @aerorahul , because the code team directly controls the documentation and it's all in the same repo.

**Comment by arunchawla-NOAA** [comment]
For NCEPLIBS-bufr and from what I can see for some of the other repos the html files have been generated and stored in gh-pages branches. Is that true ? Can we then just delete the html files loaded here ? @MinsukJi-NOAA can you confirm for me where the html files for NEMS and FV3 dycore are ?

**Reply by MinsukJi-NOAA** [reply]
NEMS doc html files are generated from [the sources](https://github.com/NOAA-EMC/NEMS/tree/develop/doc), and placed into [here](https://github.com/NOAA-EMC/NOAA-EMC.github.io)

**Reply by aerorahul** [reply]
@MinsukJi-NOAA 
Why are they placed in this repository and not in the NEMS repository?
This is not a repository for hosting documentation for the various repositories.
The documentation should stay in the repository that hosts the code that generates it and is likely to be updated in the event of a code update.

**Reply by MinsukJi-NOAA** [reply]
It is a setup that existed when I inherited from @RatkoVasic-NOAA

**Reply by arunchawla-NOAA** [reply]
Minsuk, we want to remove the html pages from this repo. Can you move these to the nems repo, following the approach used for NCEPLIBS ? Essentially the html pages would be in a branch called gh-pages

**Reply by arunchawla-NOAA** [reply]
then this repo can only contain the stuff related to the landing page

**Reply by RatkoVasic-NOAA** [reply]
@MinsukJi-NOAA do you have all permissions to do this?

**Reply by MinsukJi-NOAA** [reply]
I believe I do, Ratko. Will let you know otherwise.

**Reply by MinsukJi-NOAA** [reply]
It's done: https://github.com/NOAA-EMC/NEMS/tree/gh-pages

**Reply by aerorahul** [reply]
and here is the link to them https://noaa-emc.github.io/NEMS/

**Reply by edwardhartnett** [reply]
Each repos documentation currently points to documentation for the most recent release.

We can also support multiple releases with gh-pages, with some effort. See https://stackoverflow.com/questions/47643881/github-pages-maintaining-multiple-versions

Or this site can be used for the documentation at versions involved in a public release, while the repo gh-pages branch holds the documentation for the most recent release.

**Comment by arunchawla-NOAA** [comment]
We need to have a standard approach for everything. Then the github.io should only contain links to the landing pages

**Comment by Hang-Lei-NOAA** [comment]
Okay, I will remove the NCEPLIBS part.

**Reply by arunchawla-NOAA** [reply]
Hang, these html files were presumably created a long time ago and should be ok to be deleted right ? Can you please confirm when you have completed this ?

**Comment by WenMeng-NOAA** [comment]
The UPP doc can be found at https://upp.readthedocs.io/en/latest/

**Reply by arunchawla-NOAA** [reply]
Wen is that the only documentation ? What about the inline doxygen created code documentation ?

**Reply by arunchawla-NOAA** [reply]
these files were created 15 months ago 

https://github.com/NOAA-EMC/NOAA-EMC.github.io/tree/master/UPP/html

Does anyone know when these were created ?

**Reply by arunchawla-NOAA** [reply]
Since this doxygen related documentation was created about 15 months ago, and as far as I can see is incomplete, my suggestion would be that we delete these html pages and create the UPP html files in the correct location (a gh-pages branch in the repo) when the UPP team starts working on the doxygen documentation. Thoughts ?

**Reply by WenMeng-NOAA** [reply]
> Wen is that the only documentation ? What about the inline doxygen created code documentation ?

That's the official documentation maintained by DTC UPP team. I will ask them the status of doxygen document.

**Reply by WenMeng-NOAA** [reply]
> these files were created 15 months ago
> 
> https://github.com/NOAA-EMC/NOAA-EMC.github.io/tree/master/UPP/html
> 
> Does anyone know when these were created ?

I believe that was created by Valbona. But it seems she never brought up to us.

**Reply by EricRogers-NOAA** [reply]
Similar DTC-created documentation for the ufs-weather-model is at  https://ufs-weather-model.readthedocs.io/en/latest

As I see it the DTC documentation is more "how to set it up and run it" while the Doxygen documentation goes more into the guts of the codes themselves, defining what each module/routine in the code is for , what the input/output files are, and so forth.

**Reply by edwardhartnett** [reply]
In the case of UPP, NCEPLIBS-bufr, and UFS_UTILS, the readthedocs documentation is a high-level documentation, maintained as a separate set of markdown files. The doxygen documentation is very detailed, extending to the level of each function. So the two serve different purposes. We have links to the readthedocs documentation, when it exists, in the README, along with a link to the doxygen docs.

We have a long-term goal of combining the two, so that the readthedocs can be generated by doxygen, and are a subset of the doxygen docs. This is not an urgent goal, however.

**Reply by fossell** [reply]
Regarding UPP, @edwardhartnett summarizes the current documentations correctly. RTD is more user guide "how to" documentation. Those rst files reside within the UPP repo under docs/ and versions are created associated with release branches (for UFS app releases).  @edwardhartnett did initial effort to implemetn doxygen standards into UPP codes, but that's as far as we've gotten on that front for inline code detail documentation.  I do not know where this https://github.com/NOAA-EMC/NOAA-EMC.github.io/tree/master/UPP/html came from, this is the first I'm seeing it.

**Reply by fossell** [reply]
I apologize that I'm out of the loop on this, can someone help bring me up to speed?  What is the goal of NOAA-EMC.github.io? Just landing page with links to various code bases? Or is it specifically supposed to have detailed code documentation only? Or all documentation?  Links to the code repos?  Do we anticipate that code-level documentation from doxygen will be created from @edwardhartnett's initial work in the UPP repo and populated here on NOAA-EMC.github.io.  Perhaps Valbona's effort was just initial attempt?  Please help me understand the goal so we can help make sure all documentations are populating in correct places, without duplication.

**Reply by arunchawla-NOAA** [reply]
Kate I put my reply at the bottom. Can you take a look there ? Sorry.

**Comment by EricRogers-NOAA** [comment]
The EMC web team will be glad to help here once some specifics beyond "clean it up" are provided. What exactly do you want this new page to link to? The most recent documentation (doxygen-generated or otherwise) pages for the various apps?

**Comment by Hang-Lei-NOAA** [comment]
Yes, I will remove the NCEPLIBS ones, and confirm upon pull request to the
repo.

**Comment by arunchawla-NOAA** [comment]
For documentation to a specific tag (I am talking FV3-dycore and NEMS here) we would have to create gh-pages branches at those tags ?

**Reply by edwardhartnett** [reply]
So if you wish to keep historical versions on-line, then perhaps this repo is the best place. In which case, @Hang-Lei-NOAA should not delete these docs, because they represent the state of play when GFS 16 was created, and that's what is desired.

**Reply by arunchawla-NOAA** [reply]
Can we create a gh-pages branch at the tag level ? And then create a static link here ? We only need to worry about the releaser documentation for FV3 and NEMS. The documentation for the libraries can be removed

**Reply by edwardhartnett** [reply]
We can support multiple gh-pages releases. See https://stackoverflow.com/questions/47643881/github-pages-maintaining-multiple-versions. So yes, we can support specific tags (which will point to releases).

**Comment by edwardhartnett** [comment]
OK, I have a suggestion. Currently the NCEPLIBS displays only the most recent version of the documentation (that is, the documentation at the time of the most recent release.)

How about we modify that procedure slightly, so that we also preserve the previous versions of the documentation. That way, we will have the documentation for every release.

Then this repo can point to a specific version of the documentation, and yet users can find the most recent release, which is probably what they want to see.

**Comment by arunchawla-NOAA** [comment]
What is the proposed action plan here ? Is it a simple case of creating gh-pages branches at the specific tags (if they do not exist yet) ? One step is to clean out the html pages that are in the repository here and move them to the relevant places. The second step is to create an appropriate solution for viewing multiple versions of the documentation

**Comment by edwardhartnett** [comment]
I will work out how to maintain past versions of the docs on the gh-pages branch (see https://github.com/NOAA-EMC/NCEPLIBS-g2c/issues/126 and https://github.com/NOAA-EMC/UFS_UTILS/issues/570, where I will work out the details).

This will handle future needs. Future releases can be supported by including links here to the appropriate versions of the documentation, which will all be available.

As far as this repo is concerned, does it currently represent the documentation of the released ufs_weather_model? Or some other product? Or is it just a random collection of the software documentation? If the former, then it should be left alone. If the latter than it should be replaced with pointers to the gh-pages branches of the various repos.

**Comment by arunchawla-NOAA** [comment]
for NEMS and FV3 (tagged versions) this place contains the released documentation so this probably needs to stay here. Everything else should be removed. Regardless of that the landing page can be made better

**Comment by arunchawla-NOAA** [comment]
@fossell to help answer your question, no nothing should sit here. The UPP documentation should sit with the UPP repo. This can be a landing page to just be a collection of links (we are still trying to figure that out). Your rst documentation is correct. For doxygen stuff when doxygen is used to create html pages these should sit in a branch called gh-pages in your UPP repo (like its being done for the libraries and UFS-Utils). Putting the html pages here was a mistake and shall be removed. Others please comment

**Reply by arunchawla-NOAA** [reply]
@edwardhartnett any comments on Kate's issue ?

**Reply by fossell** [reply]
@arunchawla-NOAA - Thank you, this is clear and makes sense.  One thing we may want to consider is having different versions of doxygen documentation (e.g. develop branch is different from public-v2 branch, or other implementation branchs). I see this discussed in other comments here. We will probably need some help to set that up properly and get familiar with the procedures are we move forward with doxygen in UPP.

**Comment by edwardhartnett** [comment]
I think the best thing for the UPP documentation would be to do as we do
for NCEPLIBS, and keep it on the gh-pages branch of the repo. WIth a simple
change in repo settings, the gh-pages can be easily displayed by the UPP
software team. It should be updated as part of the release process, so that
it does not represent the UPP develop branch, but instead documents a
released version of the software.

I can help get this going, but I don't have permissions on the UPP repo. I
have an issue up there which describes the need for the gh-pages branch to
host the documentation.

Happy to help further as needed...

Ed

**Reply by arunchawla-NOAA** [reply]
one of the things we know is how documentation in rst docs can be connected to specific tags using readthedocs. It is not clear how to do that with doxygen and gh-pages. I know Ed you were going to look into it. Maybe once that works for the libraries we can set that up for UPP ?

**Reply by WenMeng-NOAA** [reply]
> I think the best thing for the UPP documentation would be to do as we do for NCEPLIBS, and keep it on the gh-pages branch of the repo. WIth a simple change in repo settings, the gh-pages can be easily displayed by the UPP software team. It should be updated as part of the release process, so that it does not represent the UPP develop branch, but instead documents a released version of the software. I can help get this going, but I don't have permissions on the UPP repo. I have an issue up there which describes the need for the gh-pages branch to host the documentation. Happy to help further as needed... Ed
> […](#)
> On Tue, Aug 24, 2021 at 9:04 AM Kate Fossell ***@***.***> wrote: @arunchawla-NOAA <https://github.com/arunchawla-NOAA> - Thank you, this is clear and makes sense. One thing we may want to consider is having different versions of doxygen documentation (e.g. develop branch is different from public-v2 branch, or other implementation branchs). I see this discussed in other comments here. We will probably need some help to set that up properly and get familiar with the procedures are we move forward with doxygen in UPP. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#17 (reply in thread)](https://github.com/NOAA-EMC/NOAA-EMC.github.io/discussions/17#discussioncomment-1227971)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJIOMMHEHU7Q5SWKFZFMUXDT6OYFFANCNFSM5CMGQCVQ> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email> .

@edwardhartnett Can you provide us a link of gh-pages branch for NCEPLIBs? Thanks!

**Reply by edwardhartnett** [reply]
For info about using gh-pages with UPP, see https://github.com/NOAA-EMC/EMC_post/issues/226

**Comment by edwardhartnett** [comment]
We already understand how to put URLs into readthedocs documentation, I
believe.

What I am looking into is generating the readthedocs documentation and the
doxygen documentation, automatically, from the doxygen source. This would
eliminate the need for the separate readthedocs high-level documentation we
are now supporting.

However, right now, we can change the readthedocs documentation to include
links the doxygen documentation on gh-pages. I believe this has been done
in the bufr library, for example.
